### PR TITLE
Painless: Single-Quoted Strings

### PR DESCRIPTION
**Pull request by jdconrad**
Added single-quoted strings to Painless + tests.
Added String -> char and char -> String casting.
Fixed a minor bug related to statically-typed shortcuts in lists.

### CODE REVIEW

**Comment by rmuir** [comment]
I'll review it. This should be a huge win.

Today we don't support `doc['field']`, leaving only `doc["field"]` and `doc.field`. We document/recommend `doc.field` only because double quote escaping such as `doc[\"field\"]` is hellacious. But `doc.field` is much slower.

Being able to support `doc['field']` gives us consistency with the script engines today, and it gives a 50% performance increase (we gave it more type information, removes dynamic lookup).

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/Utility.java [hunk 1]
```diff
@@ -248,6 +248,10 @@ public static Double charToDouble(final char value) {
         return (double)value;
     }
 
+    public static String charToString(final char value) {
+        return "" + value;
+    }
+
```
Comment (rmuir) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/Utility.java]: Can we do this one as `String.valueOf`? https://docs.oracle.com/javase/7/docs/api/java/lang/String.html#valueOf%28char%29

Comment (jdconrad) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/Utility.java]: Done.

**Comment by rmuir** [comment]
Looks good. I pushed a commit with some docs and tests changes and ran benchmarks.

**Comment by jdconrad** [comment]
Used merge --squash.  This made it into master.

**Pull request closed without merging**